### PR TITLE
Feature/idl types [18706]

### PR DESCRIPTION
**Pull request by adriancampo**
Added IDL definitions for all the types to be tested.

### CODE REVIEW

File: IDL/primitives.idl [hunk 1]
```diff
@@ -0,0 +1,48 @@
+struct Integrals
```
Comment (richiware) [inline IDL/primitives.idl]: I will split this idl file into two: `basic_types.idl` with next types:
```
<base_type_spec> ::= <integer_type>
| <floating_pt_type>
| <char_type>
| <wide_char_type>
| <boolean_type>
| <octet_type>
```

and `strings.idl`.

Also I will split struct with multiple types into multiple structs with one type. Example:
```
struct SignedShortInt
{
    short var_short;
};
```

File: IDL/maps.idl [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    bitfield<4>;
+    bitfield<10> c;
+    bitfield<12, short> d;
+};
+
+typedef short alias_array[2];
+typedef sequence<short> alias_sequence;
+typedef map<long, long> alias_map;
+
+struct Maps
```
Comment (richiware) [inline IDL/maps.idl]: I would split it in multiple structures with one map inside.

File: IDL/sequences.idl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    bitfield<10> c;
+    bitfield<12, short> d;
+};
+
+
+typedef short alias_array[2];
+typedef sequence<short> alias_sequence;
+typedef map<long, long> alias_map;
+
+struct Sequences
```
Comment (richiware) [inline IDL/sequences.idl]: I would split in multiple structures each one with one sequence.

File: IDL/sequences.idl [hunk 4]
```diff
@@ -0,0 +1,84 @@
+enum MyEnum
+{
+    ENUM_VALUE_1,
+    ENUM_VALUE_2,
+    ENUM_VALUE_3
+};
+
+@bit_bound(8)
+bitmask MyBitMask
```
Comment (richiware) [inline IDL/sequences.idl]: I join these helper types (equal in several idl files) in a outer idl file `helpers/basic_inner_types.idl`

File: IDL/arrays.idl [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+bitset MyBitset
+{
+    bitfield<3> a;
+    bitfield<1> b;
+    bitfield<4>;
+    bitfield<10> c;
+    bitfield<12, short> d;
+};
+
+struct ArraySingleDimension
```
Comment (richiware) [inline IDL/arrays.idl]: I would split in multiple structures, each one with one array.

File: IDL/aliases.idl [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+bitset MyBitset
+{
+    bitfield<3> a;
+    bitfield<1> b;
+    bitfield<4>;
+    bitfield<10> c;
+    bitfield<12, short> d;
+};
+typedef MyBitset alias_bitset;
```
Comment (richiware) [inline IDL/aliases.idl]: Which if the purpose of this idl file if there is no at least one struct containing the aliases?

File: IDL/structures.idl [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+struct MyStructureEmptyChild: MyStructure
+{
+};
+
+struct MyEmptyStructure
+{
+};
+
+struct MyEmptyStructureChild : MyEmptyStructure
```
Comment (richiware) [inline IDL/structures.idl]: The structures in `includes.idl` and these ones with inherintance, I would move it to a idl file `inheritance.idl`.

File: IDL/unions.idl [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        long long second;
+};
+
+union Union_alias switch (MyAlias)
+{
+    case 1:
+        long first;
+    case 2:
+        long long second;
+};
```
Comment (richiware) [inline IDL/unions.idl:253]: Without structures containing these unions, they cannot be tested.

**Review by richiware — CHANGES_REQUESTED** [review]

File: IDL/primitives.idl (hunk 1 above)
Comment (adriancampo) [reply IDL/primitives.idl]: split in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/maps.idl (hunk 2 above)
Comment (adriancampo) [reply IDL/maps.idl]: split in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/sequences.idl (hunk 3 above)
Comment (adriancampo) [reply IDL/sequences.idl]: split in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/sequences.idl (hunk 4 above)
Comment (adriancampo) [reply IDL/sequences.idl]: created in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/arrays.idl (hunk 5 above)
Comment (adriancampo) [reply IDL/arrays.idl]: split in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/aliases.idl (hunk 6 above)
Comment (adriancampo) [reply IDL/aliases.idl]: created in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/structures.idl (hunk 7 above)
Comment (adriancampo) [reply IDL/structures.idl]: moved in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/unions.idl (hunk 8 above)
Comment (adriancampo) [reply IDL/unions.idl:253]: created in https://github.com/eProsima/dds-types-test/pull/1/commits/7807affbc3bdb9302cd109f1e64c8a0bf837d7c8

File: IDL/maps.idl [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+struct MapKeyULongLong
+{
+    std::map<unsigned long long, unsigned long long> var_map_ulonglong;
+};
+
+struct MapKeyFloat
+{
+    std::map<float, float> var_map_float;
+};
```
Comment (JLBuenoLopez) [inline IDL/maps.idl]: From XTYPES v1.3 specification clause 7.2.2.4.3:

> Map types have an additional attribute, the key element type, that indicates the type of the key elements. Implementers of this specification need only support key elements of signed and unsigned integer types and of narrow and wide string types; the behavior of maps with other key element types is undefined and may not be portable. (Design rationale, non-normative: Support for arbitrary key element types would require implementers to provide uniform sorting and/or hashing operations, which would be impractical on many platforms. In contrast, these operations have straightforward implementations for integer and string types.)

Several map keys should be removed according to the specification  (or at least that is the minimal support required).

Comment (adriancampo) [reply IDL/maps.idl]: Removed in [e3abe74](https://github.com/eProsima/dds-types-test/pull/1/commits/e3abe748b33c5d42e70f957439410b6722e52980)

File: IDL/maps.idl [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+struct MapBitset
+{
+    map<long, MyBitset> var_map_bitset;
+};
+    
+struct MapKeyShort
+{
+    std::map<short, short> var_map_short;
```
Comment (JLBuenoLopez) [inline IDL/maps.idl]:
```suggestion
    map<short, short> var_map_short;
```

These is repeated in the following struct definitions.

**Review by JLBuenoLopez — CHANGES_REQUESTED** [review]

File: IDL/maps.idl (hunk 10 above)
Comment (adriancampo) [reply IDL/maps.idl]: My autocomplete loves to do that....
Removed in [45f5a04](https://github.com/eProsima/dds-types-test/pull/1/commits/45f5a04ea19d9a9d5c3320836c663922d0f039d9)

File: IDL/helpers/basic_inner_types.idl [hunk 11]
```diff
@@ -0,0 +1,37 @@
+enum MyEnum
```
Comment (richiware) [inline IDL/helpers/basic_inner_types.idl]: Remove `My` and use more descriptive names. ex: InnerEnumHelper

File: IDL/primitives.idl [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+{
+    long double var_longdouble;
+};
+
+struct Boolean
+{
+    boolean var_boolean;
+};
+
+struct Octect
```
Comment (richiware) [inline IDL/primitives.idl]:
```suggestion
struct Octet

```

File: IDL/unions.idl [hunk 13]
```diff
@@ -0,0 +1,428 @@
+#include "helpers/basic_inner_types.idl"
+
+typedef short alias_array[2];
```
Comment (richiware) [inline IDL/unions.idl]: Move to `basic_inner_types.idl`

File: IDL/unions.idl [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+        long first;
+    case 2:
+        long long second;
+};
+
+const boolean discriminator_true = TRUE;
+const boolean discriminator_false = FALSE;
+union Union_Discriminator_boolean switch (boolean)
+{
+    case discriminator_true:
```
Comment (richiware) [inline IDL/unions.idl]:
```suggestion
    case TRUE:

```

File: IDL/unions.idl [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+        long first;
+    case discriminator_false:
```
Comment (richiware) [inline IDL/unions.idl]:
```suggestion
    case FALSE:

```

File: IDL/unions.idl [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+        long first;
+    case 1:
+        long long second;
+};
+
+const char discriminator_a = 'a';
+const char discriminator_b = 'b';
+union Union_Discriminator_char switch (char)
+{
+    case discriminator_a:
```
Comment (richiware) [inline IDL/unions.idl]:
```suggestion
    case 'a':

```

File: IDL/unions.idl [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+        long first;
+    case discriminator_b:
```
Comment (richiware) [inline IDL/unions.idl]:
```suggestion
    case 'b':

```

File: IDL/declarations.idl [hunk 18]
```diff
@@ -0,0 +1,135 @@
+union ForwardUnion;
+struct ForwardStruct;
+
```
Comment (richiware) [inline IDL/declarations.idl:21]: create structures which uses the forwarded types before defining them.

**Review by richiware — CHANGES_REQUESTED** [review]

File: IDL/declarations.idl (hunk 18 above)
Comment (adriancampo) [reply IDL/declarations.idl:21]: moved in [14fb2cf](https://github.com/eProsima/dds-types-test/pull/1/commits/14fb2cf413071ae76adcc7ea5a7cc0f936cc5cc4)

File: IDL/unions.idl (hunk 13 above)
Comment (adriancampo) [reply IDL/unions.idl]: moved in [14fb2cf](https://github.com/eProsima/dds-types-test/pull/1/commits/14fb2cf413071ae76adcc7ea5a7cc0f936cc5cc4)

File: IDL/helpers/basic_inner_types.idl (hunk 11 above)
Comment (adriancampo) [reply IDL/helpers/basic_inner_types.idl]: changed in moved in [14fb2cf](https://github.com/eProsima/dds-types-test/pull/1/commits/14fb2cf413071ae76adcc7ea5a7cc0f936cc5cc4)

File: IDL/primitives.idl [hunk 19]
```diff
@@ -0,0 +1,64 @@
+struct ShortStructStruct
```
Comment (richiware) [inline IDL/primitives.idl:1]:
```suggestion
struct ShortStruct

```

**Review by richiware — CHANGES_REQUESTED** [review]

File: IDL/sequences.idl [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+{
+    sequence<char> var_sequence_char;
+};
+
+struct SequenceWChar
+{
+    sequence<wchar> var_sequence_wchar;
+};
+
+struct SequenceString
```
Comment (richiware) [inline IDL/sequences.idl:68]: Please, also test bounded string

**Review by richiware — CHANGES_REQUESTED** [review]

File: IDL/sequences.idl (hunk 20 above)
Comment (adriancampo) [reply IDL/sequences.idl:68]: added in [0d61a14](https://github.com/eProsima/dds-types-test/pull/1/commits/0d61a14ee6643315c71678927715d06d5eb0220c)

File: IDL/unions.idl [hunk 21]
```diff
@@ -0,0 +1,444 @@
+#include "helpers/basic_inner_types.idl"
```
Comment (richiware) [inline IDL/unions.idl:1]: I miss a union with `default` label.

**Review by richiware — CHANGES_REQUESTED** [review]

File: IDL/unions.idl (hunk 21 above)
Comment (adriancampo) [reply IDL/unions.idl:1]: Added in [80f8b5b](https://github.com/eProsima/dds-types-test/pull/1/commits/80f8b5ba871562eb71479fbbe31a51e50321b1ca)

**Review by richiware — APPROVED** [review]

**Pull request merged**